### PR TITLE
Implement CORS middleware

### PR DESCRIPTION
**Pull request by kryptt**
This starts to address issue #218

 @rossabaker please help me clean this up a bit.

### CODE REVIEW

File: server/src/main/scala/org/http4s/server/middleware/CORS.scala [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+object CORS {
+  type CORSF = Function1[Request, Task[Option[Response]]]
+  private[CORS] val logger = getLogger
+
+  def DefaultCORSConfig = CORSConfig(
+    anyOrigin= true,
+    allowCredentials= true,
+    maxAge = 1.day.toSeconds)
+
+  def apply(service: HttpService, config: CORSConfig = DefaultCORSConfig): HttpService = Service.lift(new CORS(service, config))
```
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/middleware/CORS.scala]: I've been considering currying our middlewares to make it convenient to get a HttpService => HttpService signature through partial application:

    def apply(config: CORSConfig)(service: HttpService)

It makes a lot of sense to me as a convention, except when there's an optional param with a default, like this.

Opinions welcome.

File: server/src/main/scala/org/http4s/server/middleware/CORS.scala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    case (_, Some(origin), Some(acrm)) if allowCORS(origin, acrm) =>
+      service.runT(req).map(corsHeaders(_, origin.value, acrm.value, acrh(req))).run
+    case _ => service(req)
+  })
+
+  def options(req: Request, origin: Header, acrm: Header): Task[Option[Response]] = {
+    logger.debug(s"Options request for $origin")
+    service(req).map{_.cata(
+      corsHeaders(_, origin.value, acrm.value, acrh(req)).some,
+      corsHeaders(Response(), origin.value, acrm.value, acrh(req)).some
```
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/middleware/CORS.scala]: You can DRY up the corsHeaders calls;

    service.or(req, Ok()).map { resp => corsHeaders(...) }

**Comment by bryce-anderson** [comment]
First things first: thanks for the PR. Second thing: I know little about CORS. All that said, the code looks great to me. I can take a deeper look this weekend after doing a bit of CORS reading unless someone more knowledgeable beats me to it.

File: server/src/main/scala/org/http4s/server/middleware/CORS.scala [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        acrm)),
+      Header("Access-Control-Allow-Headers", config.allowedHeaders.cata(
+        ah => ah.mkString("", " ", ""),
+        acrh))
+    )
+  }
+
+  def allowCORS(origin: Header, acrm: Header) : Boolean = (config.anyOrigin, config.anyMethod, origin.value, acrm.value) match {
+    case (true, true, _, _)           => true
+    case (true, false, _, acrm)       => config.allowedMethods.map(_.contains(acrm))   | false
```
Comment (bryce-anderson) [inline server/src/main/scala/org/http4s/server/middleware/CORS.scala]: the ```foo | false``` is interesting. I believe ```(x: Boolean) | false``` should always be ```x```, therefor its purpose must be a form of documentation?

Comment (kryptt) [inline server/src/main/scala/org/http4s/server/middleware/CORS.scala]: actually, I'm pulling the value out of an Option[Boolean]  
it's a little syntactic suger from scalaz  basically I'm doing `option.getOrElse(false)`. I just think `|` is less noisy than `getOrElse`

File: server/src/main/scala/org/http4s/server/middleware/CORS.scala (hunk 1 above)
Comment (kryptt) [inline server/src/main/scala/org/http4s/server/middleware/CORS.scala]: I played around with the signature and tried to keep that spirit...So, yes, I agree a signature of `HttpService => HttpService` helps describe middlewares perfectly.

**Comment by kryptt** [comment]
(It seems the failure is unrelated...)

**Comment by rossabaker** [comment]
@kryptt , I remember you saying there was a problem, but I can't remember if that happened before or after the latest round of commits?  Is this good to review now?

**Comment by kryptt** [comment]
Yes, its good for review now. 
I am using it without issue on my dev and staging environments right now...

**Comment by rossabaker** [comment]
:+1:

**Pull request merged**